### PR TITLE
Add documentation for how OpenCL runtime manages buffers

### LINKED ISSUE #78: Document how buffer are managed
Add documentation page in `/docs` about how buffers are created allocated, used during kernel execution.

### PR DESCRIPTION
**Pull request by sherry-yuan**
It specifically documents;
1. how it is used
2. how sycl use it
3. some caveats
4. how memory are allocated
5. device address encoding
6. the whole implementation flow of all function that are responsible for allocation, memory transfer
7. some corner cases

Closes #78

### CODE REVIEW

File: docs/buffers.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+
+Some other note about memory operation during enqueue:
+1. Device local pointer size is 4, device global pointer size is always the device's address bit integer division 8
+
+Before [submitting kernel](https://github.com/intel/fpga-runtime-for-opencl/blob/1264543c0361530f5883e35dc0c9d48ac0fd3653/src/acl_kernel.cpp#L2982-L3093) to device queue. It first checks if the device is programmed, if not, it will [queue reprogram device operation](https://github.com/intel/fpga-runtime-for-opencl/blob/1264543c0361530f5883e35dc0c9d48ac0fd3653/src/acl_kernel.cpp#L3034) to do so. Then it will [arrange memory migration](https://github.com/intel/fpga-runtime-for-opencl/blob/1264543c0361530f5883e35dc0c9d48ac0fd3653/src/acl_kernel.cpp#L3043) for each kernel memory argument.
+
+Different from the device operation resulting from enqueue read/write, the memory migration calls on [`acl_mem_migrate_buffer`](https://github.com/intel/fpga-runtime-for-opencl/blob/b08e0af97351718ce0368a9ee507242b35f4929e/src/acl_mem.cpp#L5412-L5665) (i.e memory transfer and memory migration behave differently). 
+
+In memory migration:
```
Comment (aditikum) [inline docs/buffers.md:132]: shouldn't this also be a section ?

**Review by aditikum — APPROVED** [review]
Thanks for creating this documentation!

File: docs/buffers.md (hunk 1 above)
Comment (sherry-yuan) [reply docs/buffers.md]: Thanks for the approval! I have wrapped memory migration into a subsection.

**Comment by sherry-yuan** [comment]
@aditikum - Thanks for the approval! The approval are dismissed after new commits, you may need to re-approve.

@vijethamalkai @pcolberg @zibaiwan @huangbri - Does this seem ok to you?

**Review by zibaiwan — APPROVED** [review]
@sherry-yuan , sorry for the late appoval. Your documentation looks great to me. Thanks for adding this. It's very helpful for runtime development.

**Comment by sherry-yuan** [comment]
Thanks @vijethamalkai for the revision and suggestions! The latest commit [0879ed8](https://github.com/intel/fpga-runtime-for-opencl/pull/81/commits/0879ed8c9634927f63a5cb4d798d65e7d23449fb) address Vij's suggestion on the use of ambiguous 'it' 'this' 'that' etc.

File: docs/buffers.md [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+### `clSetKernelArg`: What if the `clEnqueueWriteBuffer` function is not called?
+
+When the `clEnqueueWriteBuffer` function is not called, memory transfer automatically happens before launching the kernel that uses it. There is an enqueued memory transfer device operation before every kernel launch device operation. The only difference between calling and not calling the `enqueueWriteBuffer` function is whether the enqueued memory transfer actually copies the memory.
```
Comment (pcolberg) [inline docs/buffers.md:107]: that uses it → that uses the buffer

File: docs/buffers.md [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+* Device address has a special encoding, which is the same between buffer and USM pointers.
+* Even if you do not explicitly transfer the memory before launching the kernel, the memory is still migrated right before the kernel executes (i.e., `clEnqueueWriteBuffer` is not necessary).
+
+## SYCL Runtime Flow
+
+SYCL* runtime calls the memory operation in the following order, which are explained in the subsections below:
+1. `clCreateBuffer`
+2. `clEnqueueWriteBuffer`
+3. `clSetKernelArg`
+4. `clEnqueueKernel` / Task
```
Comment (pcolberg) [inline docs/buffers.md:57]:
```suggestion
4. `clEnqueueKernel` / `clEnqueueTask`
```

File: docs/buffers.md [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+// Submit kernel execution to queue
+err = clEnqueueNDRangeKernel(queue, kernel, 2, NULL, &global_work_size_2d[0], &local_work_size_2d[0], 0, NULL, &event);
+
+... // Block on kernel finish and clean up resources
+```
+
+## Summary
+
+* Buffers are not allocated on the device until the kernel executes, the `clEnqueueWriteBuffer` function is called, or `CL_MEM_COPY_HOST_PTR` flag is defined when calling the function `clCreateBuffer`.
```
Comment (pcolberg) [inline docs/buffers.md:43]: Swapping the order preserves the flow:
```suggestion
* Buffers are not allocated on the device until the kernel executes, the `clEnqueueWriteBuffer` function is called, or the `clCreateBuffer` function is called with the `CL_MEM_COPY_HOST_PTR` flag.
```

File: docs/buffers.md [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+* Buffer handles memory migration between the host and device automatically.
+* SYCL* runtime typically calls OpenCL functions in a different order in comparison to OpenCL's preferred order. Specifically, SYCL* runtime calls functions in the order of `clCreateBuffer` > `clEnqueueWriteBuffer` > `clSetKernelArg` > `clEnqueueNDRangeKernel`. In OpenCL, the optimal calls are `clCreateBuffer` > `clSetKernelArg` > `clEnqueueWriteBuffer` >  `clEnqueueNDRangeKernel`. The difference in order of calls sometimes leads to an unexpected result (eg. extra host to device memory copy), but you can avoid it if the `clSetKernelArg` function does not modify `cl_mem` attributes.
```
Comment (pcolberg) [inline docs/buffers.md:45]: Does the latter preference actually apply to OpenCL and (most of) its implementations, or only to the FPGA Runtime for OpenCL?

File: docs/buffers.md [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+## SYCL Runtime Flow
+
+SYCL* runtime calls the memory operation in the following order, which are explained in the subsections below:
+1. `clCreateBuffer`
+2. `clEnqueueWriteBuffer`
+3. `clSetKernelArg`
+4. `clEnqueueKernel` / Task
+
+
+**NOTE:** SYCL* runtime calls are not the preferred order from the OpenCL runtime perspective. However, SYCL runtime* cannot change the order of API invocations as it needs to handle other vendors' requirements. Not all vendors can get away with setting kernel argument first before enqueuing write buffer (As discussed in one of the [issues](https://github.com/intel/llvm/discussions/4627)).
```
Comment (pcolberg) [inline docs/buffers.md:60]: OpenCL runtime → FPGA Runtime for OpenCL

File: docs/buffers.md (hunk 5 above)
Comment (pcolberg) [reply docs/buffers.md:45]: This item is quite long and the order of calls cumbersome to read. How about removing the entire sentence "Specifically …" in favour of a link to the section below? Then expand the section below to include a second list of calls in the order preferred by the FPGA Runtime for OpenCL.

File: docs/buffers.md (hunk 6 above)
Comment (pcolberg) [reply docs/buffers.md:60]: (As discussed in one of the [issues](https://github.com/intel/llvm/discussions/4627)) → (see [discussion](https://github.com/intel/llvm/discussions/4627))

**Review by pcolberg — CHANGES_REQUESTED** [review]
Thanks @sherry-yuan for the great work, and thanks @vijethamalkai for the corrections :slightly_smiling_face:

File: docs/buffers.md [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+cl_kernel kernel = clCreateKernel(program, "kernel_name", &err);
+
+// Host side data that contains the information needed for kernel execution
+unsigned int *host_data = (unsigned int *)acl_aligned_malloc(global_work_size*sizeof(unsigned int)*GMEM_DATA_VEC);
+
+// Set up the expectation for a buffer, but not actually allocating memory yet
+cl_mem mem = clCreateBuffer(context, CL_MEM_READ_WRITE, global_work_size*sizeof(unsigned int)*GMEM_DATA_VEC, NULL, &err);
+
+// Enqueue a memory transfer operation
+cl_event event = {};
```
[the remark this answers is not in the document]
Comment (sherry-yuan) [reply docs/buffers.md]: Thank you for all the suggestions!! I have made the changes accordingly to other places as well.

**Review by pcolberg — APPROVED** [review]

**Pull request merged**